File: 1022.SumofRootToLeafBinaryNumbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumRootToLeaf = function(root) {
    const stack = [[root, '']];
    const result = [];
    while(stack.length) {
        const [node, ancestor] = stack.shift();
        if (node) {
            const temp = ancestor + node.val;
            if (node.left) {
                stack.push([node.left, temp]);
            } 
            if (node.right) {
                stack.push([node.right, temp]);
            }
            if (!node.left && !node.right) {
                result.push(temp);
            }
        }
    }
    return result.reduce((acc, item) => acc + parseInt(item, 2), 0);
};

File: Golang/Tree/1022.Sum-of-root-to-leaf-binary-numbers.go
package main

import (
	"fmt"
	"strconv"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumRootToLeaf(root *TreeNode) int {
	type TreeNodePath struct {
		Node *TreeNode
		Path string
	}
	paths := []string{}
	stack := []*TreeNodePath{&TreeNodePath{
		Node: root,
		Path: "",
	}}
	for len(stack) > 0 {
		child := []*TreeNodePath{}
		for _, s := range stack {
			newPath := fmt.Sprintf("%s%d", s.Path, s.Node.Val)
			if s.Node.Left == nil && s.Node.Right == nil {
				paths = append(paths, newPath)
			}
			if s.Node.Left != nil {
				child = append(child, &TreeNodePath{
					Node: s.Node.Left,
					Path: newPath,
				})
			}
			if s.Node.Right != nil {
				child = append(child, &TreeNodePath{
					Node: s.Node.Right,
					Path: newPath,
				})
			}
		}
		stack = child
	}
	var ret int64 = 0
	for _, path := range paths {
		val, _ := strconv.ParseInt(path, 2, 32)
		ret += val
	}
	return int(ret)
}
